criteo_distributed_training/utils.py

```diff
@@ -0,0 +1,135 @@
+import argparse
+import os
+import ray
+
+from thirdai import bolt
+
+
+def cpus_per_node_type(value):
+    # 2 is here for testing purpose, as we might not want to start whole cluster
+    valid_values = [2, 12, 24, 48]
+    if int(value) not in valid_values:
+        raise argparse.ArgumentTypeError(
+            f"Invalid value for cpus_per_node: {value}. Valid values are {valid_values}"
+        )
+    return int(value)
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        description="Distributed training with Ray for Universal Deep Transformer (UDT) model"
+    )
+    parser.add_argument(
+        "--embedding_dimension",
+        type=int,
+        default=256,
+        metavar="N",
+        help="Embedding dimension for the UDT model (default: 256)",
+    )
+    parser.add_argument(
+        "--num_nodes",
+        type=cpus_per_node_type,
+        required=True,
+        default=2,
+        metavar="N",
+        help="Number of CPUs allocated per node for the distributed training. Valid values: 2, 12, 24, 48 (default: 12)",
+    )
+    parser.add_argument(
+        "--cpus_per_node",
+        type=int,
+        default=4,
+        metavar="N",
+        help="Number of CPUs allocated per node for the distributed training (default: 4)",
+    )
+    parser.add_argument(
+        "--epochs",
+        type=int,
+        default=1,
+        metavar="N",
+        help="Number of epochs to train the model (default: 1)",
+    )
+    parser.add_argument(
+        "--learning_rate",
+        type=float,
+        default=0.005,
+        metavar="LR",
+        help="Learning rate for the model (default: 0.005)",
+    )
+    parser.add_argument(
+        "--batch_size",
+        type=int,
+        default=720000,
+        metavar="N",
+        help="Batch size for training (default: 720000)",
+    )
+    parser.add_argument(
+        "--max_in_memory_batches",
+        type=int,
+        default=3,
+        metavar="N",
+        help="Maximum number of in-memory batches (default: 10)",
+    )
+    args = parser.parse_args()
+    return args
+
+
+def setup_ray(num_nodes=2, cpus_per_node=4):
+    working_dir = os.path.dirname(os.path.realpath(__file__))
+
+    ray.init(
+        runtime_env={
+            "working_dir": working_dir,
+            "env_vars": {
+                "OMP_NUM_THREADS": f"{cpus_per_node}",
+                "GLOO_SOCKET_IFNAME": "ens5",
+            },
+            "excludes": ["trained_models", "test_file.txt"],
+        },
+        ignore_reinit_error=True,
+    )
+    scaling_config = ray.air.ScalingConfig(
+        num_workers=num_nodes,
+        use_gpu=False,
+        trainer_resources={"CPU": 16},
```

Comment:
Can we increase it to 24, right? Since the head node has 72 CPUs and we would be using 48 CPUs only.

Reply:
Changed. Earlier I was using a different machine which had 64 cpus but have now shifted to 72 cpus machine.